src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -15,29 +14,17 @@ internal sealed class JsonPropertyInfoNotNullable<TClass, TDeclaredProperty, TRu
         JsonPropertyInfoCommon<TClass, TDeclaredProperty, TRuntimeProperty>
         where TRuntimeProperty : TDeclaredProperty
     {
-        // Constructor used for internal identifiers
-        public JsonPropertyInfoNotNullable() { }
-
-        public JsonPropertyInfoNotNullable(
-            Type parentClassType,
-            Type declaredPropertyType,
-            Type runtimePropertyType,
-            PropertyInfo propertyInfo,
-            Type elementType,
-            JsonSerializerOptions options) :
-            base(parentClassType, declaredPropertyType, runtimePropertyType, propertyInfo, elementType, options)
-        {
-        }
-
         public override void Read(JsonTokenType tokenType, JsonSerializerOptions options, ref ReadStack state, ref Utf8JsonReader reader)
         {
+            Debug.Assert(ShouldDeserialize);
+
             if (ElementClassInfo != null)
             {
                 // Forward the setter to the value-based JsonPropertyInfo.
                 JsonPropertyInfo propertyInfo = ElementClassInfo.GetPolicyProperty();
                 propertyInfo.ReadEnumerable(tokenType, options, ref state, ref reader);
             }
-            else if (ShouldDeserialize)
+            else
```

Comment:
I still think we should avoid such deep nesting here:
https://github.com/dotnet/corefx/pull/37578#discussion_r283076814

Reply:
I commented on the link - the code is also trying to prevent pre-initialization of `value`.